### PR TITLE
Add Go 1.23 support, bump to Go 1.22.6, drop Go 1.21 support

### LINKED ISSUE #198: Update to Go Version 1.23
**Issue or Feature Request:**

Go released version 1.23.0 on August 13 https://go.dev/doc/devel/release#go1.23.0. Update bottlerocket-sdk to support Go 1.23 and drop Go 1.21 support per Go's support policy https://go.dev/doc/devel/release#policy

### PR DESCRIPTION
**Pull request by ginglis13**
**Issue number:**

Closes #198 

**Description of changes:**

Go 1.23.0 and 1.22.6 have been released. Following the [Go release policy](https://go.dev/doc/devel/release#policy), drop support for Go 1.21. Set the default Go major version to 1.23

There is now a minor diff between `patches/go-1.22/0002-Always-restrict-boringcrypto-crypto-tls-to-FIPS.patch` and `patches/go-1.23/0002-Always-restrict-boringcrypto-crypto-tls-to-FIPS.patch` as the import statement for `go/src/crypto/tls/boring.go` was moved to one line.

**Testing done:**

- [x] Build SDK `x86_64`
- [x] Build SDK `aarch64`
- [x] Build Bottlerocket `aarch64` on `x86_64`
- [x] Build Bottlerocket `aarch64` on `aarch64`
- [x] Build Bottlerocket `x86_64` on `x86_64`
- [x] Build Bottlerocket `x86_64` on `aarch64`

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: patches/go-1.23/0002-Always-restrict-boringcrypto-crypto-tls-to-FIPS.patch [hunk 1]
```diff
@@ -0,0 +1,42 @@
+From 5256a813afbb9c0f7d7ae00544ae1cbaeafb7f1e Mon Sep 17 00:00:00 2001
+From: Gavin Inglis <giinglis@amazon.com>
+Date: Tue, 13 Aug 2024 22:00:50 +0000
+Subject: [PATCH] Always restrict boringcrypto crypto/tls to FIPS
+
+Signed-off-by: Gavin Inglis <giinglis@amazon.com>
```
Comment (bcressey) [inline patches/go-1.23/0002-Always-restrict-boringcrypto-crypto-tls-to-FIPS.patch]: I don't feel strongly about this, but for a patch that you didn't create but need to adjust context for an update, you can follow this convention from kernel development:
```suggestion
From 5256a813afbb9c0f7d7ae00544ae1cbaeafb7f1e Mon Sep 17 00:00:00 2001
From: Gavin Inglis <giinglis@amazon.com>
Date: Tue, 13 Aug 2024 22:00:50 +0000
Subject: [PATCH] Always restrict boringcrypto crypto/tls to FIPS

Signed-off-by: Ben Cressey <bcressey@amazon.com>
[giinglis: update for Go 1.23]
Signed-off-by: Gavin Inglis <giinglis@amazon.com>
```

Goal is to retain the original authorship but also take credit / blame for the changes you made.

**Review by bcressey — APPROVED** [review]

File: patches/go-1.23/0002-Always-restrict-boringcrypto-crypto-tls-to-FIPS.patch (hunk 1 above)
Comment (ginglis13) [reply patches/go-1.23/0002-Always-restrict-boringcrypto-crypto-tls-to-FIPS.patch]: That's definitely a more appropriate outcome and reminiscent of coauthored / cherry picked commits. I'll adjust 👍

**Comment by ginglis13** [comment]
^ force push adjusts `patches/go-1.23/0002-Always-restrict-boringcrypto-crypto-tls-to-FIPS.patch` to retain bcressey's credits

**Comment by ginglis13** [comment]
Smoke testing complete and ready for review

**Review by bcressey — APPROVED** [review]

**Review by jpculp — APPROVED** [review]

**Comment by ginglis13** [comment]
I notice in the release notes https://go.dev/doc/go1.23

> Setting the GOROOT_FINAL environment variable no longer has an effect ([#62047](https://go.dev/issue/62047)). Distributions that install the go command to a location other than $GOROOT/bin/go should install a symlink instead of relocating or copying the go binary.

We set that here: https://github.com/bottlerocket-os/bottlerocket-sdk/blob/b21e50c7f1318d44e44b8e1fca37eec90ccd273b/helpers/go/build-go.sh#L15

But I do have go 1.23.0 installed

```
docker run public.ecr.aws/o7o0w6s5/thar-be-a-beta-sdk:v0.44.0 go version
go version go1.23.0 linux/arm64
```

and it does exist under /usr/libexec:

```
docker run public.ecr.aws/o7o0w6s5/thar-be-a-beta-sdk:v0.44.0 ls /usr/libexec
...
go-1.22
go-1.23
...
```

This is likely because the `go` provided by `fedora:39`, our base image, is still `go 1.22.5`:

```
$ docker run public.ecr.aws/docker/library/fedora:39 dnf -y install go && go version
...
Complete!
go version go1.22.5 linux/arm6
```

I created an issue to track this, since we might be burned by this when (if?) fedora:39 updates to go 1.23.x: #200

**Pull request merged**